src/fsharp/TypeChecker.fs

```diff
@@ -5067,7 +5067,18 @@ and RecordNameAndTypeResolutions_IdeallyWithoutHavingOtherEffects_Delayed cenv e
             RecordNameAndTypeResolutions_IdeallyWithoutHavingOtherEffects cenv env tpenv arg
             dummyCheckedDelayed otherDelayed
         | _ -> ()
-    dummyCheckedDelayed delayed 
+
+    dummyCheckedDelayed delayed
+
+// Calls UnifyTypes, records name resolutions on failure, so that type information
+// can be collected for IntelliSense.
+and UnifyTypes_RecordNameResolutionsOnError cenv env m expectedTy actualTy tpenv delayed =
+    try
+        UnifyTypes cenv env m expectedTy actualTy
+    with
+        _ ->
+            RecordNameAndTypeResolutions_IdeallyWithoutHavingOtherEffects_Delayed cenv env tpenv delayed
```

Comment:
This unfortunately hurts performance, but only in case an error occurs in `UnifyTypes`.